### PR TITLE
Add support for subgrids and fix some other grid properties

### PR DESCRIPTION
**Pull request by ghost**
This covers the `subgrid <line-name-lists>?` for `grid-template` (for gecko, and hence, stylo).

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17268)

### CODE REVIEW

File: components/style/values/generics/grid.rs [hunk 1]
```diff
@@ -556,3 +557,93 @@ impl<T: ToCss> ToCss for TrackList<T> {
...
+
+/// The `<line-name-list>` for subgrids.
+///
+/// `[ <line-names> | repeat(<positive-integer> | auto-fill, <line-names>+) ]+`
+/// **NOTE; This is deprecated!**
+#[derive(Clone, PartialEq, Debug)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub struct LineNameList {
+    /// The optional `<line-name-list>`
+    pub names: Vec<Vec<CustomIdent>>,
```
Comment (nox) [inline components/style/values/generics/grid.rs]: I wonder if that should be `Box<[Box<[CustomIdent]>>`, but that's unrelated to that patch.

File: components/style/values/generics/grid.rs [hunk 2]
```diff
@@ -556,3 +557,93 @@ impl<T: ToCss> ToCss for TrackList<T> {
         Ok(())
     }
 }
+
+/// The `<line-name-list>` for subgrids.
+///
+/// `[ <line-names> | repeat(<positive-integer> | auto-fill, <line-names>+) ]+`
+/// **NOTE; This is deprecated!**
```
Comment (nox) [inline components/style/values/generics/grid.rs]: How is this deprecated? Didn't it just got moved to a different greater version of the spec?

File: components/style/values/generics/grid.rs [hunk 3]
```diff
@@ -556,3 +557,93 @@ impl<T: ToCss> ToCss for TrackList<T> {
         Ok(())
     }
 }
+
+/// The `<line-name-list>` for subgrids.
```
Comment (nox) [inline components/style/values/generics/grid.rs]: Spec link would be nice.

File: components/style/values/specified/grid.rs [hunk 4]
```diff
@@ -349,3 +349,52 @@ impl ToComputedValue for TrackList<TrackSizeOrRepeat> {
...
+    #[inline]
+    fn has_viewport_percentage(&self) -> bool {
+        match *self {
+            GridTemplateComponent::TrackList(ref l) => l.has_viewport_percentage(),
+            _ => false,
+        }
+    }
+}
+
+impl ToComputedValue for GridTemplateComponent<TrackSizeOrRepeat> {
```
Comment (nox) [inline components/style/values/specified/grid.rs]: This can be derived.

**Review by nox — CHANGES_REQUESTED** [review]

**Comment by wafflespeanut** [comment]
I'll address the comments once I'm convinced about the test failures.

**Comment by wafflespeanut** [comment]
@canaltinova I just talked to @bholley. I won't be very active for at least a month (except in IRC and responding to comments). Could you take ownership of this PR? Feel free to ping me for anything that doesn't make sense (or look weird). Also, this was my last try run - https://treeherder.mozilla.org/#/jobs?repo=try&revision=2385382ec93c8f6dc480b5b06c1cecd25c6e1d48&selectedJob=108086299

**Comment by canaltinova** [comment]
@wafflespeanut Sure, I'll work on that next.

**Comment by canaltinova** [comment]
Latest try job: https://treeherder.mozilla.org/#/jobs?repo=try&revision=76c0777c8719950bccc2647e233544da9cdf131b We still need to fix the remaining 53 failures. But this looks like a good start. I think we can land this and deal with the rest later.

**Comment by canaltinova** [comment]
@bors-servo r=wafflespeanut,canaltinova
(Oops, I thought that test can accept 2 parameters somehow.)

**Comment by canaltinova** [comment]
@bors-servo r=wafflespeanut,canaltinova
(more test changes)

**Pull request merged**